File: DSA/DP/LongestCommonSubsequence.py
class Solution:
    def longestCommonSubsequence(self, s1: str, s2: str) -> int:
        m = len(s1)
        n = len(s2)
         
        t = [[-1 for j in range(n+1)] for i in range(m+1)]
        
        for i in range(m+1):
            t[i][0] = 0
        for j in range(n+1):
            t[0][j] = 0
        
        for i in range(1, m+1):
            for j in range(1, n+1):
                if s1[i-1] == s2[j-1]:
                    t[i][j] = 1 + t[i-1][j-1]
                else:
                    t[i][j] = max(t[i-1][j], t[i][j-1])
        return t[m][n]
        

File: Java/longestCommonSubsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int n= text1.length();
        int m=text2.length();
        int[][]dp=new int[n+1][m+1];
          for(int i=0;i<=n;i++)
              dp[i][0]=0;
          
         for(int j=0;j<=m;j++)
             dp[0][j]=0;
        
        for(int i=1;i<=n;i++){
        for(int j=1;j<=m;j++){
          if(text1.charAt(i-1)!=text2.charAt(j-1))
                 dp[i][j]= Math.max(dp[i][j-1],dp[i-1][j] );
             else dp[i][j]=1+dp[i-1][j-1];
         }
        }
        return dp[n][m];
    }
}  
